File: typescript/word_counter/word_counter_12.ts
/**
 * This program is a word counter written in TypeScript.
 * It reads a text input, processes the text to count the number of words,
 * and caches the data in memory for quick access. The program is designed
 * to be verbose and well-commented to help programmers understand each step.
 */

// Import necessary modules
import * as fs from 'fs';

// Function to read text from a file
function readTextFile(filePath: string): string {
    // Read the file synchronously
    const data = fs.readFileSync(filePath, 'utf8');
    return data;
}

// Function to count words in a given text
function countWords(text: string): number {
    // Split the text by spaces and filter out empty strings
    const words = text.split(/\s+/).filter(word => word.length > 0);
    return words.length;
}

// Function to cache data in memory
let cache: { [key: string]: number } = {};

function cacheWordCount(filePath: string, wordCount: number): void {
    // Store the word count in the cache
    cache[filePath] = wordCount;
}

// Main function to execute the word counting process
function main(filePath: string): void {
    // Read the text from the file
    const text = readTextFile(filePath);

    // Count the number of words in the text
    const wordCount = countWords(text);

    // Cache the word count in memory
    cacheWordCount(filePath, wordCount);

    // Output the word count to the console
    console.log(`The file ${filePath} contains ${wordCount} words.`);
}

// Execute the main function with a sample file path
main('sample.txt');

File: ruby/word_counter/word_counter_12.rb
# This program is a word counter written in Ruby.
# It reads a text file, counts the occurrences of each word,
# and prints the word counts in descending order.
# The program also caches data in memory to improve performance.

# Method to read the file and return its content
def read_file(file_path)
    # Open the file in read mode
    file = File.open(file_path, "r")
    # Read the entire content of the file
    content = file.read
    # Close the file
    file.close
    # Return the content
    content
end

# Method to count the occurrences of each word in the text
def count_words(text)
    # Split the text into words using spaces and punctuation as delimiters
    words = text.split(/\W+/)
    # Initialize an empty hash to store word counts
    word_counts = Hash.new(0)
    # Iterate over each word in the array
    words.each do |word|
        # Convert the word to lowercase to ensure case-insensitive counting
        word = word.downcase
        # Increment the count for the word in the hash
        word_counts[word] += 1
    end
    # Return the hash containing word counts
    word_counts
end

# Method to print the word counts in descending order
def print_word_counts(word_counts)
    # Sort the hash by values (word counts) in descending order
    sorted_word_counts = word_counts.sort_by { |word, count| -count }
    # Iterate over each key-value pair in the sorted hash
    sorted_word_counts.each do |word, count|
        # Print the word and its count
        puts "#{word}: #{count}"
    end
end

# Main program execution starts here
if __FILE__ == $0
    # Define the path to the text file
    file_path = "sample.txt"
    # Read the content of the file
    text = read_file(file_path)
    # Count the occurrences of each word in the text
    word_counts = count_words(text)
    # Cache the word counts in memory for future use
    @cached_word_counts = word_counts
    # Print the word counts
    print_word_counts(word_counts)
end

